traceapp/app.go

```diff
@@ -85,6 +88,23 @@ func (a *App) serveRoot(w http.ResponseWriter, r *http.Request) error {
 func (a *App) serveTrace(w http.ResponseWriter, r *http.Request) error {
 	v := mux.Vars(r)
 
+	if permalink := r.URL.Query().Get("permalink"); permalink != "" {
+		// If the user specified a permalink, then decode it directly into a
+		// trace structure and place it into storage for viewing.
+		gz, err := gzip.NewReader(base64.NewDecoder(base64.RawURLEncoding, strings.NewReader(permalink)))
```

Comment:
If you're going to bypass URL query encoding (this is akin to using package `unsafe`, or using `text/template` to generate HTML), I highly recommend being familiar with how it works.

The end result looks valid to me. `base64.RawURLEncoding` consists only of characters that don't require escaping, so skipping `url.QueryEscape` step is fine since it's guaranteed to be a noop.

Specifically, see https://gotools.org/net/url#url.go-L98-L101 and https://gotools.org/net/url#url.go-L120-L121, alphanumeric and `-`, `_` characters all don't need escaping.